internal/metrics/recorder.go

```diff
@@ -330,19 +332,20 @@ func NewReconcileError(
 }
 
 // Reports a reconcile error as a prometheus metric. The parameter crName
-// is the name of the CR being reconciled by the controller.
+// is the name of the CR being reconciled by the controller and returns the
+// processed error.
 func (r *ReconcileError) Report(err error, crName string) {
 	if r.recorder == nil {
 		return
 	}
-	newErr := err.Error()
-	// Retrieve the specific subreconciler error if present
-	if r.isSubReconcilerError {
-		if unwrapped := errors.Unwrap(err); unwrapped != nil {
-			newErr = unwrapped.Error()
-		}
+	var ctlrReconcileErr *controllers.ControllerReconcileError
+	// Errors can be arbitrarily returned anywhere in the reconcile loop.
+	// To avoid collecting arbitrary/unclassified ones a check for its type
+	// is necessary.
+	if !errors.As(err, &ctlrReconcileErr) {
+		return
```

Comment:
This kinda throws me off, while I understand the intent is to simply skip the metric recording, this seems to happen later in the process (i.e; when a controller decides to report a metric, makes the function call, and the functional call exits without relaying information). 

Is there a better strategy where we can parse the errors even before they invoke the record func and stops from recording the metric. Should happen in the controllers' code.

Reply:
> this seems to happen later in the process (i.e; when a controller decides to report a metric, makes the function call, and the functional call exits without relaying information).

I get what you are concerned about. However, this is instrumentation code (metric collection) just like how one would treat logging, the controller should not care about its result, therefore I would side on simplicity by making the controller unaware about.

>Is there a better strategy where we can parse the errors even before they invoke the record func and stops from recording the metric. Should happen in the controllers' code.

Same as above, this is instrumentation code and there is no point handling whatever input/output the Recording will process/result to. If this was part of business logic (e.g. http error check) then the Report func should not run silently.